apig/generate.go

```diff
@@ -386,131 +387,152 @@ func generateDB(detail *Detail, outDir string) error {
 	return nil
 }
 
-func Generate(outDir, modelDir, targetFile string, all bool) int {
-	outModelDir := filepath.Join(outDir, modelDir)
+func generateCommonFiles(detail *Detail, outDir string) error {
+	var wg sync.WaitGroup
+	errCh := make(chan error, 1)
+	done := make(chan bool, 1)
+
+	for _, model := range detail.Models {
+		wg.Add(1)
+		go func(m *Model) {
+			defer wg.Done()
+			d := &Detail{
+				Model:     m,
+				ImportDir: detail.ImportDir,
+				VCS:       detail.VCS,
+				User:      detail.User,
+				Project:   detail.Project,
+			}
+
+			if err := generateApibModel(d, outDir); err != nil {
+				fmt.Fprintln(os.Stderr, err)
+				errCh <- err
+			}
+
+			if err := generateController(d, outDir); err != nil {
+				fmt.Fprintln(os.Stderr, err)
+				errCh <- err
+			}
+		}(model)
+	}
+
+	wg.Wait()
+	close(done)
+
+	select {
+	case <-done:
+	case err := <-errCh:
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func collectModels(outModelDir string) (Models, error) {
 	files, err := ioutil.ReadDir(outModelDir)
 	if err != nil {
-		fmt.Fprintln(os.Stderr, err)
-		return 1
+		return nil, err
 	}
 
 	var models Models
 	var wg sync.WaitGroup
-	modelMap := make(map[string]*Model)
-	errCh := make(chan error)
-	modelsCh := make(chan []*Model)
-	doneCh := make(chan struct{})
-	go func() {
```

Comment:
In this one goroutine, other multiple goroutines are called internally. It may not make great improvement in performance.